src/cmd/ksh93/bltins/cd_pwd.c

```diff
@@ -243,13 +253,27 @@ int	b_cd(int argc, char *argv[],Shbltin_t *context)
 		path_pwd(shp,0);
 		if(*shp->pwd != '/')
 		{
-			errormsg(SH_DICT,ERROR_system(1),e_direct);
+			errormsg(SH_DICT,ERROR_system(ret),e_direct);
 			UNREACHABLE();
 		}
 	}
 	nv_scan(sh_subtracktree(1),rehash,(void*)0,NV_TAGGED,NV_TAGGED);
 	path_newdir(shp,shp->pathlist);
 	path_newdir(shp,shp->cdpathlist);
+	if(pflag && eflag)
+	{
+		/* If the directory was changed and the PWD cannot be determined,
+		   exit with status 1. */
+		char *verify_pwd = getcwd(NIL(char*),0);
+		if(!verify_pwd)
```

Comment:
I think this test is insufficient. The new option aims to guarantee that `$PWD` is correct, but that is not guaranteed just because `getcwd` succeeds. After checking that, you should also check whether the value returned refers to the same directory as that in `PWD`. This can be done with `test_inode(verify_pwd,nv_getval(pwdnod))`. That function is also used for the `-ef` test/[[ operator and it's already an extern.